proposals/1113_ldap-ad-web-ui/README.md

```diff
@@ -0,0 +1,136 @@
+When a client other than the Pinniped CLI makes an authorization endpoint request with
+custom Username/Password headers, they should be rejected.
+
+The pinniped cli should default to using the cli-based password flow, but when a tty is unavailable or when
+a CLI flag (`--ldap-browser-login`? `--force-browser-login`?) is specified, it will open a browser to log in
+instead of prompting for username and password. Some users (for example, IDE plugins for kubernetes)
+may wish to authenticate using the pinniped cli but without access to a terminal.
+
+#### Upgrades
+
+This change is backwards compatible. Users would see no changes unless they decided to register
+a new client or change the pinniped cli flags.
+
+However if they do choose to register a new client they may need to update the following:
+- FederationDomains today may be using private certificate authorities. These are trusted
+  for our use case but a browser will flag them as unsafe. Admins will have to transition to letsencrypt
+  or another public Certificate Authority to prevent making end users click past messages about the certificate
+  being untrusted.
+- The name of the idp custom resource is currently not published to users logging in with Pinniped.
+  We plan on exposing this to indicate to users which idp they are logging in to.
+  Admins may need to update this to something more user-friendly.
+
+#### Tests
+
+Chromedriver browser based integration tests will be needed to ensure that a user can log in from a web-based app
+by entering their ldap credentials into the web page.
+
+#### New Dependencies
+This should be kept to a very simple HTML page with minimal, clean CSS styling.
+Javascript should be avoided.
+
+#### Observability Considerations
+* Logging login attempts at higher log levels.
+
+#### Security Considerations
+* Preventing LDAP injection attacks: this should be done server-side using our existing
+  string escaping.
+* CSRF protection via a CSRF cookie: this should be similar to the way it is done for the
+  OIDCIdentityProvider today
+* The new UI page must be HTTPS.
+
+#### Documentation Considerations
+This new feature will require documentation to explain how to configure it and to publicise that it is available.
+This should include:
+* A blog post describing the feature
+* Website documentation in the form of a how-to guide
+
+### Other Approaches Considered
+Today, users can configure Dex if they want a web-based LDAP login.
+This introduces complexity because they have to install, configure and
+maintain both Pinniped and Dex in order to use this feature. It also means
+that users do not benefit from the opinionated `ActiveDirectoryIdentityProvider`
+config because Dex does not have an equivalent.
+
+## Open Questions
+* Is the ldap login page hosted per FederationDomain?
+* What is the format for the URL? (`issuer/some/path`? Something else?) Can we make it so we can reuse the existing cert,
+  or will we need a new wildcard cert?
```

Comment:
Requiring a new cert will make this excessively painful for admins.

Reply:
`<issuer_path>/ldap/<ldap_idp_name>/login` was suggested in https://hackmd.io/@pinniped/Hk2WA1sb_#Possible-Future-Feature-Web-Based-LDAP-Login-Support, I think that makes sense and would not require new certs.

Reply:
This URL is not an API right?  i.e. we redirect to it, and we can change that at any point (the starting point is always the authorize endpoint).  Thus I think `<issuer_path>/login` is sufficient?

Reply:
Putting the IDP name and type in the path was a suggestion to help get ready to handle the future feature of multiple IDP support. The information does not necessarily need to be in the path if there is some other way for the authorize endpoint to communicate the requested IDP name and type through the redirect to the handler for the LDAP/AD login page. The login page will need to be able to use the name/type to lookup the IDP, for example to show its name on the page.

Reply:
Related thought: If the IDP name/type is encoded into a signed/encrypted state param (rather than the path), then we can rest assured that nobody can try to directly access this endpoint without going through the authorize endpoint first. That would help us avoid needing to gracefully handle error edge cases like someone directly trying to access the new login endpoint to log into an OIDCIdentityProvider (not supported), since that would be effectively impossible because the authorize endpoint would never send them to the login endpoint for an OIDCIdentityProvider.

Reply:
I think this conversation was resolved and the doc was updated accordingly.